From b598c859e93f2189cc0bb42420d0ce069ab01bdc Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Date: Thu, 18 Oct 2012 13:06:27 +0300
Subject: [PATCH 055/162] dpaa_eth: Fix memcopy from invalid page

The memcopy of short frames into the linear part of the skb was
mistakenly being done after the source page had been released. This
occasionally resulted in corrupted TCP ACKs and possibly similarly
corrupted small ingress datagrams, visible via netstat.

Moving the memcopy where the source page is still valid.

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
[Kevin: Original patch taken from FSL
QorIQ-SDK-V1.3-SOURCE-20121114-yocto.iso tarball.]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/freescale/dpa/dpaa_eth_sg.c |    7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/freescale/dpa/dpaa_eth_sg.c b/drivers/net/ethernet/freescale/dpa/dpaa_eth_sg.c
index 4a3fb7a..31a6118 100644
--- a/drivers/net/ethernet/freescale/dpa/dpaa_eth_sg.c
+++ b/drivers/net/ethernet/freescale/dpa/dpaa_eth_sg.c
@@ -308,8 +308,8 @@ static int __hot contig_fd_to_skb(const struct dpa_priv_s *priv,
 
 	tailptr = skb_put(skb, copy_size);
 
-	/* Try to optimize the memcpy to follow */
-	prefetch(vaddr + dpa_fd_offset(fd));
+	/* Copy (at least) the headers in the linear portion */
+	memcpy(tailptr, vaddr + dpa_fd_offset(fd), copy_size);
 
 	/*
 	 * If frame is longer than the amount we copy in the linear
@@ -329,9 +329,6 @@ static int __hot contig_fd_to_skb(const struct dpa_priv_s *priv,
 		dpa_bp_add_page(dpa_bp, (unsigned long)vaddr);
 	}
 
-	/* Copy (at least) the headers in the linear portion */
-	memcpy(tailptr, vaddr + dpa_fd_offset(fd), copy_size);
-
 	return 0;
 }
 
-- 
1.7.9.7

